src/control/variable_header/protocol_level.rs

```diff
@@ -5,20 +5,27 @@ use std::io::{Read, Write};
 
 use byteorder::{ReadBytesExt, WriteBytesExt};
 
-use crate::{Decodable, Encodable};
 use crate::control::variable_header::VariableHeaderError;
+use crate::{Decodable, Encodable};
 
+pub const SPEC_3_1_0: u8 = 0x03;
 pub const SPEC_3_1_1: u8 = 0x04;
+pub const SPEC_5_0: u8 = 0x05;
 
 /// Protocol level in MQTT (`0x04` in v3.1.1)
 #[derive(Debug, Eq, PartialEq, Copy, Clone)]
-pub struct ProtocolLevel(pub u8);
+#[repr(u8)]
+pub enum ProtocolLevel {
+    Version310 = SPEC_3_1_0,
+    Version311 = SPEC_3_1_1,
+    Version50 = SPEC_5_0,
+}
```

Comment:
We shouldn't make the protocol level fixed as an `enum`, it would be nice to define some helper functions like:

```rust
impl ProtocolLevel {
    pub fn protocol_3() -> ProtocolLevel {
        ProtocolLevel(0x03)
    }
}
```

already good enough.

Reply:
From trying to write a broker I find it's more ergonomic to not even expose those inner const values-- I don't care what the inner u8 is for those and trying to remember which is which gets confusing. I just need to know what protocol version is specified (the number of legal values is very small, and they unfortunately aren't all in a single spec document). It's unfortunately not trivial because the `u8` doesn't even map neatly to the MQTT version-- spec 3.1.1 is 0x03, and for some reason they skipped spec 4 and said spec 5.0.0 is ProtocolVersion 0x04 (?) So without this parsing into an enum I'd have to do some sort of lookup by const values. It's just interpretation that I feel works best within this crate rather than the broker.

I could add `#[non_exhaustive]` to indicate that there may exist more variants than we expose, but in practice virtually every broker and client right now are either 3.1.1 or 5.0.

Reply:
> and for some reason they skipped spec 4 and said spec 5.0.0 is ProtocolVersion 0x04 (?)

https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html#_Toc385349227

Nope. v5.0 is `0x05`.

Reply:
With a bit of a chuckle at my own mistake, I think you just made my point. I don't want to think about this stuff, and the field is essentially a de-facto enumeration of repr u8. Can we just parse it into an enum? It'll make my MQTT broker project a lot easier to write.